### PR TITLE
glibc: 2.34-115 -> 2.35-30

### PR DESCRIPTION
**Pull request by Ma27**
###### Description of changes

Announcement: https://sourceware.org/pipermail/libc-alpha/2022-February/136040.html

__Note:__ *very* early stage, probably needs a Hydra jobset (cc @vcunat @jonringer )

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
let me see if hydra-unstable builds on master...

I can setup a personal jobset..... and it failed :(

**Comment by vcunat** [comment]
So, do you want a jobset on hydra.nixos.org now?  aarch64-linux is relatively free, and that most likely won't change soon.

**Comment by Mic92** [comment]
Let me know, when you drop the static-pie flag and I can test if glibc still builds pie crt files.

**Comment by vcunat** [comment]
They even say the flag was removed:
> * The --enable-static-pie option is no longer available.  The glibc build
  configuration script now automatically detects static-pie support in the
  toolchain and architecture and enables it if available.

**Comment by vcunat** [comment]
I assume that this won't aim for NixOS 22.05.

**Comment by Ma27** [comment]
> I assume that this won't aim for NixOS 22.05.

Absolutely not! The earliest for this is 22.11 IMHO.

> So, do you want a jobset on hydra.nixos.org now?

I assume that this is currently not a good idea considering that we're releasing in the foreseeable future, right?

I don't think I'll be able to do much anyways during the next weeks so I'm fine with waiting until 22.05 was released and then we can create a jobset.
Alternatively, I'd ask @jonringer  (and perhaps help if their Hydra is still broken) because I don't have the resources for that.

**Comment by vcunat** [comment]
At this point rules don't allow inclusion in 22.05 anymore.

Right now Hydra is down, but this week it had some spare x86_64-linux capacity.  We can always prioritize (cancel) in case we accumulate a too long queue.

**Comment by Ma27** [comment]
> At this point rules don't allow inclusion in 22.05 anymore.

Well aware of that ;-)

> Right now Hydra is down, but this week it had some spare x86_64-linux capacity

if that's possible currently, I'd say let's go for it :)

**Comment by kvtb** [comment]
Please add
```
disallowedReferences = lib.optional (stdenv.cc.libc != null) stdenv.cc.libc;
```

this new `glibc` refers to bootstrap's `glibc` increasing closure size

it may need
```
      patchelf --set-interpreter $(readlink -f $bin/bin/ld.so)  $bin/bin/ldconfig
      patchelf --set-interpreter $(readlink -f $bin/bin/ld.so)  $bin/bin/sln
```
in `postInstall` to fix (but I tested vanilla `2.35`,  `2.35-30` might have it fixed; anyway, `disallowedReferences` would be good to avoid future regression)

File: pkgs/development/libraries/glibc/common.nix [hunk 1]
```diff
@@ -62,14 +62,14 @@ stdenv.mkDerivation ({
...
+          $ git fetch --all -p && git checkout origin/release/2.35/master && git describe
+          glibc-2.35-30-g9be62976af
+          $ git show --minimal --reverse glibc-2.35.. | gzip -9n --rsyncable - > 2.35-master.patch.gz
 
          To compare the archive contents zdiff can be used.
-          $ zdiff -u 2.34-master.patch.gz ../nixpkgs/pkgs/development/libraries/glibc/2.34-master.patch.gz
+          $ zdiff -u 2.35-master.patch.gz ../nixpkgs/pkgs/development/libraries/glibc/2.35-master.patch.gz
        */
-      ./2.34-master.patch.gz
+      ./2.35-master.patch.gz
```
Comment (kvtb) [inline pkgs/development/libraries/glibc/common.nix:73]: why not `fetchpatch { url = "https://github.com/bminor/glibc/compare/glibc-2.35..9be62976af.patch"; hash = ...; }` ?

File: pkgs/development/libraries/glibc/common.nix [hunk 2]
```diff
@@ -62,14 +62,14 @@ stdenv.mkDerivation ({
...
```
Comment (kvtb) [inline pkgs/development/libraries/glibc/common.nix:72]: why not `fetchpatch { url = "https://github.com/bminor/glibc/compare/glibc-2.35..9be62976af.patch"; hash = ...; }` ?

Comment (vcunat) [reply pkgs/development/libraries/glibc/common.nix:72]: That's similar "would complicate bootstrapping" part, I think.  Maybe it wouldn't be too much, but so far we've been using this approach.

Comment (ajs124) [reply pkgs/development/libraries/glibc/common.nix:72]: Doesn't work. Can't `fetchpatch` while bootstrapping.

Comment (kvtb) [reply pkgs/development/libraries/glibc/common.nix:72]: `fetchurl` then

**Comment by kvtb** [comment]
also resulted `${glibc.bin}/bin/ldd` is a shell script starting with `!#/bin/bash`

it should be replaced with `!#/bin/sh` or `patchShebangs`'ed

File: pkgs/development/libraries/glibc/common.nix (hunk 2 above)
Comment (vcunat) [reply pkgs/development/libraries/glibc/common.nix:72]: That won't give a stable hash IIRC>

**Comment by ncfavier** [comment]
Do we still need the C.UTF-8 stuff in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/glibc/default.nix ?

**Comment by vcunat** [comment]
Probably not; the support should've been upstreamed, perhaps even earlier than 2.35.

**Pull request closed without merging**